### ISSUE TITLE
Creating `leave_1study_out`

### ISSUE THREAD

**Issue by itchyshin** [daniel1noble/orchaRd#10]
@daniel1noble 

I am making a note that we should add leave_1study_out akin to metafor's `leave1out`

https://rdrr.io/cran/metafor/man/leave1out.html

**Comment by fdecunta** [comment]
Hi,  

I’m not sure if you’re still into this, but I’ve been writing some functions to help me running and plotting leave-one-outs. They still need a lot of work, but I’d love to get your feedback if you’re interested!  

I uploaded some examples in this [notebook](https://rpubs.com/fdecunta/leave-one-out-min). So far, I’ve implemented:  

#### **leave_one_out(model, group, vcalc)**  
Refits a metafor model iteratively, removing each level of a grouping variable. Returns a data frame ready for plotting.  [Link to code](https://github.com/fdecunta/orchaRd/blob/feature-leave1out/R/leave_one_out.R)

#### **loo_plot(model, loo_dataframe, order, labels)**  
Plots all overall effect sizes with confidence intervals from the leave-one-out analysis, along with the original model’s effect size and CI.  [Link to code](https://github.com/fdecunta/orchaRd/blob/feature-leave1out/R/loo_plot.R)

---

Currently, it only works for intercepts, but I plan to add support for plotting other parameters, as well as adding phylogenetic and non-phylogenetic components to the models.  



**P.S.:** Thanks for creating orchaRd!

**Comment by daniel1noble** [comment]
Hi @fdecunta,

This is still on the cards, ye. What you have looks like a good start but it would be good to wrap it up as an orchard_plot to be consistent with the general structure of the package. Otherwise, it would be great to have it. @itchyshin and I are quite time crunched so these things take time for us so appreciate your help.

Cheers,
Dan

**Comment by fdecunta** [comment]
Hi Daniel,
Thanks! I’ll work on it and update you as soon as I have something to share.
Best,
Facundo

**Comment by fdecunta** [comment]
Hi Daniel,
I've being working to make this consistent with the rest of the package. If you have some time, I would like to hear your thoughts! You can see some examples in this [notebook](https://rpubs.com/fdecunta/leave-one-out-min2).

I rewrote the function as a wrapper around `orchard_plot()`. For now, it's named `orchard_leave1out()`, and it uses the same arguments as `orchard_plot()`, plus a few optional specific arguments.

Internally, the function runs all the models, store the results using `mod_results()`, and then passes them to `orchard_plot()`.  I like this approach because it integrates seamlessly with the existing codebase, requiring no changes elsewhere.

Also, it is user-friendly, like the rest of the package. For example, this is enough for a basic plot: 
```r
orchard_leave1out(res, group = "authors", xlab = "lnRR")
```

There's still more work to do, but I’d appreciate your feedback to see if you think this is the right direction.

Best,
Facundo

**Comment by daniel1noble** [comment]
Hi @fdecunta

Sorry to be slow. Very busy. I had a look at this and it looks great! One thing I can’t see for sure is, are the data subsetted and platted too? It looks like in some plots that the whole data is re-plotted, but would we not want to have the subset of data plotted without the data of the study left out? May be it’s already doing that and I just can’t tell from the plot. 

Either way, the format and structure are all really nicely consistent with orchard more generally. When we have the examples adding some tests to make sure thing are working would be good. I can add some of your example code to the vignette. For the colour blind, I agree that cb argument with limited colours may need to be turned off for these plots, but that should be pretty easy to do, or simply have a call to first ask how many colours will exist and if over the limit default to viridis (which I like)

Thanks for all your efforts here!

Cheers,
Dab

**Comment by fdecunta** [comment]
Hi @daniel1noble 

I'm glad to hear that you liked it!

Regarding the plotted points, they are a subset that excludes the left-out study. You are right that it's hard to notice. I was thinking of plotting the left-out points with a red circle, or something like that. I'll try this and see if it's worth it. 

Adding this to the vignette sounds great! I'll write some tests to ensure everything is working properly.

For the colors, I like the idea of first checking the number of colors and defaulting to viridis if needed.

Finally, I'll write a separate function for the transformations and include a couple of tests for it. 

I think that after these changes the function should be usable for simple cases. I'll work on all of this and let you know once it's ready!

Best,
Facundo

**Comment by itchyshin** [comment]
@fdecunta - I wanted to think you for all the coding you are doing on this one and other functions - thank you!

**Comment by daniel1noble** [comment]
I second what @itchyshin says! Thanks so much for helping with this @fdecunta. It's hard to maintain everything at a good rate while being so busy so it awesome to have someone like you helping out.

**Comment by fdecunta** [comment]
@itchyshin  @daniel1noble Thank you both! I've learned a lot from reading some of your papers, so it's really great to be able to contribute and help out!

**Comment by fdecunta** [comment]
Hi!
I have a new demo to show you. Here is the [notebook](https://rpubs.com/fdecunta/leave-one-out-demo). As before, the two basic functions are `orchard_leave1out()` and `leave_one_out()`. The notebook has some examples and brief explanations of the functions and their parameters. 

I've also written tests for both functions (and also for most of their auxiliary functions), so I think they should be safe for basic usage. If you want to inspect the code, you can find it in this [branch](https://github.com/fdecunta/orchaRd/tree/feature-leave1out) .

Let me know your thoughts! I’m happy to make any changes if needed.

Best,
Facundo

**Comment by itchyshin** [comment]
@fdecunta - I very much like `orchard_leave1out` - this is something I never thought about. It looks really good and visualizing which data points are taken out. As you know, this works for leave-1species-out, which we started to do and maybe more important than leave-1study-out biologically

**Comment by fdecunta** [comment]
Thanks @itchyshin ! I'm glad you liked it! As you noted, the function allows any grouping variable, so it is easy to use it for leave-1species-out. I should add an example using species instead of studies.

One problem with the function right now is that it doesn't support the use of phylogenetic correlation matrix yet. I will work on it, maybe it's not too hard to implement!

**Comment by daniel1noble** [comment]
@fdecunta Yes, agree with @itchyshin. This looks great! thanks so much for building this function. Yep, I can see why the phylogeny is tricky. I do have a function that can prune the tree quickly so that could be done within the function as you re-fit the model. So, we can include that into the package and that could solve that problem?

**Comment by daniel1noble** [comment]
Here it is below. It does more than prune, it also does some checks. We can adopt elements of it maybe?

`#' @title tree_checks
#' @description Checks and prunes trees
#' @param data Dataframe with species names
#' @param tree Phylogenetic tree
#' @param dataCol Column name of species in data
#' @param type Character. "checks" will return a list of species in the tree but not in the data and vice versa. "prune" will prune the tree of species not in the data.
  tree_checks <- function(data, tree, dataCol, type = c("checks", "prune")){
    type = match.arg(type)
    # How many unique species exist in data and tree
    Numbers <- matrix(nrow = 2, ncol = 1)
    Numbers[1,1] <- length(unique(data[,dataCol])) 
    Numbers[2,1] <- length(tree$tip.label) 
    rownames(Numbers)<- c("Species in data:", "Species in tree:")
    # Missing species or species not spelt correct      
    species_list1= setdiff(sort(tree$tip.label), sort(unique(data[,dataCol])))
    species_list2= setdiff(sort(unique(data[,dataCol])), sort(tree$tip.label) )
    if(type == "checks"){
      return(list(SpeciesNumbers = data.frame(Numbers), 
                  Species_InTree_But_NotData=species_list1, 
                  Species_InData_But_NotTree=species_list2))
    }
    if(type == "prune"){
      if(length(species_list2) >=1) stop("Sorry, you can only prune a tree when you have no taxa existing in the data that are not in the tree")
      return(ape::drop.tip(tree, species_list1))
    }
  }`

**Comment by fdecunta** [comment]
Thanks @daniel1noble , that seems to solve the problem I have in mind! I'll try to use it inside the leave-one-out function. 

I'll also try directly 'subsetting' the phylo matrix. I'm not sure which approach will be simpler (and faster), but I think we can have a working solution relatively quickly.

I'll update you once I have news!

**Comment by fdecunta** [comment]
Hi!

I have a new version that works with phylogeny. [Here](https://rpubs.com/fdecunta/leave-1spp-out) is an example. 

The user pass a list with the tree and the name of the random factor used in the model that is linked to the phylo matrix. Internally, the function uses a simplified version of Daniel's code to prune the tree and then recalculates the phylogenetic correlation matrix in each run. 

I’ve written tests for each new function and also ran some manual tests. Seems to be working fine, but take a look at the notebook and let me know if you spot any problem or have any questions!

**Comment by itchyshin** [comment]
@fdecunta - fabulous! - thank you

**Comment by daniel1noble** [comment]
Looks fantastic! Does this also work if, say, you’re still interested in leaving 1 study out but have fit the model with a phylogeny. In that case, you still might have species drop out. That would make it a little more general as I think more people Ould have a phylogeny but want to leave 1 study out not species. Anyway, this is all looking great and thanks so much for all your wonderful efforts!

**Comment by fdecunta** [comment]
Great, I'm really glad you liked it! 
@daniel1noble you're right, the function can leave out elements from any group while still using the phylogeny in the model. I just tried and it worked!

I'll write a nice documentation with examples, and play a little bit with the functions to see what problems do I find.

Once the documentation is ready, feel free to use the code if you’d like! I can create a pull request or, if you don't want to add it to your repo, I will maintain it in my fork. 

Either way, it was awesome to have your feedback. Thank you so much!

**Comment by daniel1noble** [comment]
Hi Facundo,

I had a look. I think this looks great! I love how it looks and functions. I’d say create a pull request and I can merge into the main branch. I can also add your little example to the vignette and update that too. 

Very cool stuff. Thank you so much for all your effort on this! 

Cheers,
Dan

**Comment by fdecunta** [comment]
Hi Dan,

Thanks for your kind words! I'm really happy you liked it.

I just created the pull request. Let me know if any issues come up.

I really enjoyed working on this and learned a lot in the process! If there’s anything else I can help with, now or in the future, just let me know. You can find my email in my profile.

Thanks again!

Cheers,
Facundo

**Comment by itchyshin** [comment]
Thanks @fdecunta - a great addition! - I mention it on bluesky - https://bsky.app/profile/itchyshin.bsky.social/post/3lmchqfiugc2e

**Comment by fdecunta** [comment]
Thank you very much @itchyshin ! That's so cool!

**Comment by daniel1noble** [comment]
@fdecunta Found some time to update the vignette with the example now and also one that introduces this new plot. All finished and update pushed. Thank you so much again for your work here. I made explicit reference to you in the vignette to make it clear this was all you!

Cheers,
Dan

**Comment by fdecunta** [comment]
Thanks a lot, @daniel1noble ! I I really appreciate the kind acknowledgment in the vignette. Happy to have contributed!
